api/providers/tasks.py

```diff
@@ -563,10 +582,96 @@ def handle_registration_row(row, initiator, provider, schema, auto_approval=Fals
             registration.sanction.accept()
         except Exception as e:
             raise RegistrationBulkCreationRowError(
-                row.upload.id, row.id, row_title, row_external_id,
+                row.upload.id, row.id, title, external_id,
                 error=repr(e), approval_failure=True,
             )
         logger.info(f'Registration approved but pending moderation: [{registration._id}]')
+    return registration
+
+
+def bulk_upload_finish_job(upload, row_count, success_count, draft_errors, approval_errors, dry_run=True):
+    """Finish the bulk upload job: handle errors (if any), send outcome emails, and update job state.
+    """
+    provider = upload.provider
+    auto_approval = upload.provider.bulk_upload_auto_approval
+    initiator = upload.initiator
+
+    # All registration rows have failed draft creation
+    if len(draft_errors) == row_count:
+        upload.state = JobState.DONE_ERROR
+        message = f'All registration rows failed during bulk creation. ' \
+                  f'Upload ID: [{upload.id}], Draft Errors: [{draft_errors}]'
+        sentry.log_message(message)
+        logger.error(message)
+    # Some registration rows have failed draft creation
+    elif draft_errors:
+        upload.state = JobState.DONE_PARTIAL
+        if not approval_errors:
+            message = f'Some registration rows failed during bulk creation. ' \
+                      f'Upload ID: [{upload.id}]; Draft Errors: [{draft_errors}].'
+        else:
+            message = f'Some registration rows failed during bulk creation and some failed to be approved. ' \
+                      f'Upload ID: [{upload.id}]; Draft Errors: [{draft_errors}]; Approval Errors: [{approval_errors}]'
```

Comment:
There is a rare case that approval/accept fail due to elastic-search down. The registration has already been created (i.e. registered and approval required). Thus, `approval_errors` is logged but does not affect the "success/failure" state fo the bulk upload job.